packages/compiler-cli/ngcc/src/host/umd_host.ts

```diff
@@ -547,51 +551,216 @@ function getUmdWrapper(statement: ts.Statement):
   return null;
 }
 
+/**
+ * Parse the wrapper function of a UMD module and extract info about the factory function calls for
+ * the various formats (CommonJS, AMD, global).
+ *
+ * The supported format for the UMD wrapper function body is a single statement which is a
+ * `ts.ConditionalExpression` (i.e. using a ternary operator). For example:
+ *
+ * ```js
+ * // Using a conditional expression:
+ * (function (global, factory) {
+ *   typeof exports === 'object' && typeof module !== 'undefined' ?
+ *     // CommonJS2 factory call.
+ *     factory(exports, require('foo'), require('bar')) :
+ *   typeof define === 'function' && define.amd ?
+ *     // AMD factory call.
+ *     define(['exports', 'foo', 'bar'], factory) :
+ *     // Global factory call.
+ *     (factory((global['my-lib'] = {}), global.foo, global.bar));
+ * }(this, (function (exports, foo, bar) {
+ *   // ...
+ * }));
+ * ```
+ */
+function parseUmdWrapperFunction(wrapperFn: ts.FunctionExpression): UmdModule['factoryCalls'] {
+  const stmt = wrapperFn.body.statements[0];
+  let conditionalFactoryCalls: UmdConditionalFactoryCall[];
+
+  if (ts.isExpressionStatement(stmt) && ts.isConditionalExpression(stmt.expression)) {
+    conditionalFactoryCalls = extractFactoryCallsFromConditionalExpression(stmt.expression);
+  } else {
+    throw new Error(
+        'UMD wrapper body is not in a supported format (expected a conditional expression):\n' +
+        wrapperFn.body.getText());
+  }
+
+  const factoryCalls = {
+    amdDefine: getAmdDefineCall(conditionalFactoryCalls),
+    commonJs: getCommonJsFactoryCall(conditionalFactoryCalls),
+    global: getGlobalFactoryCall(conditionalFactoryCalls),
+  };
+
+  if (factoryCalls.commonJs === null) {
+    throw new Error(
+        'Unable to find a CommonJS factory call inside the UMD wrapper function:\n' +
+        stmt.getText());
+  }
+
+  return factoryCalls as (typeof factoryCalls&{commonJs: ts.CallExpression});
```

Comment:
Would it make sense to do:

```suggestion
  const commonJs = getCommonJsFactoryCall(conditionalFactoryCalls);
   if (commonJs === null) {
     throw new Error(
         'Unable to find a CommonJS factory call inside the UMD wrapper function:\n' +
         stmt.getText());
   }

   return {
     amdDefine: getAmdDefineCall(conditionalFactoryCalls),
     commonJs,
     global: getGlobalFactoryCall(conditionalFactoryCalls),
   };
```

that avoids the slightly awkward cast in the return position.

**EDIT** Ah, the subsequent commit may explain why it's done like this; perhaps the above structure doesn't apply that nicely in that case anymore.

Reply:
Yeah, this might work for one property, but not for two (as we need in a subsequent commit).
I couldn't find a way to implement this in a way that TS understands that "at least one of two properties is non-null" (without type-casting) 😞